pyat/at.c

```diff
@@ -291,6 +291,23 @@ void set_energy_particle(PyObject *lattice, PyObject *energy,
             }
             Py_DECREF(particle);
          }
+          PyObject *bcurrent = PyObject_GetAttrString(lattice, "beam_current");
+          PyArrayObject *bspos = (PyArrayObject *) PyObject_GetAttrString(lattice, "_bunch_spos");
+          PyArrayObject *fpattern = (PyArrayObject *) PyObject_GetAttrString(lattice, "_fillpattern");
+          if (bcurrent != NULL) {
+              param->beam_current = PyFloat_AsDouble(bcurrent);
+              Py_DECREF(bcurrent);
+          }
+          if (bspos != NULL && fpattern != NULL) {
+              param->bunch_spos = PyArray_DATA(bspos);
+              param->nbunch = (int)PyArray_SIZE(bspos);
+              param->bunch_currents = malloc(param->nbunch*sizeof(double));
+              double *fill_pattern = PyArray_DATA(fpattern);
+              for(int i=0; i<param->nbunch; i++){
+                  param->bunch_currents[i]=param->beam_current*fill_pattern[i];
+              }
+              Py_DECREF(fpattern);
```

Comment:
This gives a memory leak:  `param->bunch_currents` is not released.

The easier solution is to do the multiplication in python (in `lattice_pass`), and send the preprocessed bunch currents in a additional keyword argument. The question of Py_DECREF also disappears.

Reply:
I did not like so much to change ``lattice_pass`` so I integrated a new attribute ``_bunch_currents`` to ``Lattice``. A private method updates it when needed.
This should fix memory issues. As far as I understood ``PyObject_GetAttrString`` returns a borrowed reference, and ``PyArray_DATA`` has no effect on reference counting, could you confirm?

Reply:
Did you check this?